File: frontend/e2e/export.ts
import { crc32 } from "node:zlib";

/**
 * A Letterboxd account export, built in the test, for the import journey to upload.
 *
 * The archive is written by hand rather than through a zip library, because the smoke
 * suite exists to prove the stack is wired together and is not worth a dependency. Every
 * member is stored uncompressed, which is a legal zip and the simplest one to emit: a
 * local header per file, a central directory, and the end record.
 */

const HEADER = ["Date", "Name", "Year", "Letterboxd URI"] as const;

export interface Row {
  name: string;
  year: number;
  rating?: number;
}

/** The five files that matter, with the headers a real export was verified to carry. */
export function letterboxdExport(ratings: Row[], watchlist: Row[]): Buffer {
  return zip([
    ["ratings.csv", csv([...HEADER, "Rating"], ratings.map(line))],
    ["watchlist.csv", csv([...HEADER], watchlist.map(line))],
    ["watched.csv", csv([...HEADER], ratings.map(line))],
    [
      "diary.csv",
      csv(
        [...HEADER, "Rating", "Rewatch", "Tags", "Watched Date"],
        ratings.map((row) => [...line(row), "", "", "2024-05-01"]),
      ),
    ],
    ["profile.csv", csv(["Username", "Email Address", "Favorite Films"], [["owner", "", ""]])],
  ]);
}

function line(row: Row): string[] {
  return [
    "2024-05-01",
    row.name,
    String(row.year),
    `https://boxd.it/${row.name.replace(/\W/g, "").toLowerCase()}`,
    row.rating === undefined ? "" : String(row.rating),
  ];
}

/** Quoted the way Letterboxd quotes: a comma inside a title must survive the round trip. */
function csv(header: readonly string[], rows: string[][]): string {
  const cell = (value: string) => (/[",\n]/.test(value) ? `"${value.replace(/"/g, '""')}"` : value);
  const width = header.length;
  return [header, ...rows]
    .map((row) => row.slice(0, width).map(cell).join(","))
    .join("\n")
    .concat("\n");
}

// --- The archive ---

function zip(members: [string, string][]): Buffer {
  const locals: Buffer[] = [];
  const central: Buffer[] = [];
  let offset = 0;

  for (const [name, body] of members) {
    const filename = Buffer.from(name, "utf8");
    const content = Buffer.from(body, "utf8");
    const sum = crc32(content);

    const local = Buffer.alloc(30);
    local.writeUInt32LE(0x04034b50, 0); // local file header
    local.writeUInt16LE(20, 4); // version needed
    local.writeUInt16LE(0, 6); // flags
    local.writeUInt16LE(0, 8); // stored, not deflated
    local.writeUInt32LE(0, 10); // no timestamp worth inventing
    local.writeUInt32LE(sum, 14);
    local.writeUInt32LE(content.length, 18); // compressed size
    local.writeUInt32LE(content.length, 22); // uncompressed size
    local.writeUInt16LE(filename.length, 26);
    local.writeUInt16LE(0, 28); // no extra field
    locals.push(local, filename, content);

    const entry = Buffer.alloc(46);
    entry.writeUInt32LE(0x02014b50, 0); // central directory header
    entry.writeUInt16LE(20, 4); // version made by
    entry.writeUInt16LE(20, 6); // version needed
    entry.writeUInt16LE(0, 8);
    entry.writeUInt16LE(0, 10);
    entry.writeUInt32LE(0, 12);
    entry.writeUInt32LE(sum, 16);
    entry.writeUInt32LE(content.length, 20);
    entry.writeUInt32LE(content.length, 24);
    entry.writeUInt16LE(filename.length, 28);
    entry.writeUInt16LE(0, 30); // extra
    entry.writeUInt16LE(0, 32); // comment
    entry.writeUInt16LE(0, 34); // disk number
    entry.writeUInt16LE(0, 36); // internal attributes
    entry.writeUInt32LE(0, 38); // external attributes
    entry.writeUInt32LE(offset, 42);
    central.push(entry, filename);

    offset += local.length + filename.length + content.length;
  }

  const directory = Buffer.concat(central);
  const end = Buffer.alloc(22);
  end.writeUInt32LE(0x06054b50, 0); // end of central directory
  end.writeUInt16LE(0, 4);
  end.writeUInt16LE(0, 6);
  end.writeUInt16LE(members.length, 8);
  end.writeUInt16LE(members.length, 10);
  end.writeUInt32LE(directory.length, 12);
  end.writeUInt32LE(offset, 16);
  end.writeUInt16LE(0, 20); // no archive comment
  return Buffer.concat([...locals, directory, end]);
}

File: backend/tests/export.py
"""Letterboxd account exports, built in memory, for the seed import's tests.

The headers are the ones verified against a real 592-row export (#16), written through
``csv`` rather than joined by hand, so a comma inside a title is quoted exactly as
Letterboxd quotes it and a naive parser fails here the way it would fail on the real
file.

``extras`` writes the rest of the archive the importer must never read: reviews and
comments at the root, the likes folder, and the deleted and orphaned folders whose
``diary.csv`` looks exactly like the one that does count.
"""

import csv
import io
import zipfile
from dataclasses import dataclass

NAME = "letterboxd-owner-2026-08-02-11-00-utc.zip"

EMAIL = "owner@example.com"
"""profile.csv's PII: nothing in the account realm may ever come to hold it."""


@dataclass(frozen=True)
class Row:
    """One exported line. The defaults are a plain, unremarkable film."""

    name: str
    year: int | None = 2010
    rating: float | None = None
    uri: str | None = None
    date: str = "2024-05-01"
    watched_date: str | None = None
    rewatch: bool = False

    @property
    def year_text(self) -> str:
        return "" if self.year is None else str(self.year)

    @property
    def rating_text(self) -> str:
        """Whole stars serialise without a decimal, exactly as Letterboxd writes them."""
        if self.rating is None:
            return ""
        return str(int(self.rating)) if self.rating == int(self.rating) else str(self.rating)

    @property
    def uri_text(self) -> str:
        return self.uri or f"https://boxd.it/{abs(hash(self.name)) % 100000:05x}"


def export(
    *,
    ratings: tuple[Row, ...] = (),
    watchlist: tuple[Row, ...] = (),
    watched: tuple[Row, ...] = (),
    diary: tuple[Row, ...] = (),
    favorites: tuple[str, ...] = (),
    extras: bool = True,
    omit: tuple[str, ...] = (),
) -> bytes:
    """A zip of the five files that matter, plus whatever else a real export carries."""
    members = {
        "ratings.csv": _csv(
            ("Date", "Name", "Year", "Letterboxd URI", "Rating"),
            [(r.date, r.name, r.year_text, r.uri_text, r.rating_text) for r in ratings],
        ),
        "watchlist.csv": _csv(
            ("Date", "Name", "Year", "Letterboxd URI"),
            [(r.date, r.name, r.year_text, r.uri_text) for r in watchlist],
        ),
        "watched.csv": _csv(
            ("Date", "Name", "Year", "Letterboxd URI"),
            [(r.date, r.name, r.year_text, r.uri_text) for r in watched],
        ),
        "diary.csv": _diary(diary),
        "profile.csv": _profile(favorites),
    }
    if extras:
        members.update(_discarded())
    return _zip({name: body for name, body in members.items() if name not in omit})


def _diary(rows: tuple[Row, ...]) -> str:
    return _csv(
        ("Date", "Name", "Year", "Letterboxd URI", "Rating", "Rewatch", "Tags", "Watched Date"),
        [
            (
                row.date,
                row.name,
                row.year_text,
                row.uri_text,
                row.rating_text,
                "Yes" if row.rewatch else "",
                "",
                row.watched_date or row.date,
            )
            for row in rows
        ],
    )


def _profile(favorites: tuple[str, ...]) -> str:
    """The whole PII header, so a parser reading more than one column is caught here."""
    return _csv(
        (
            "Date Joined",
            "Username",
            "Given Name",
            "Family Name",
            "Email Address",
            "Location",
            "Website",
            "Bio",
            "Pronoun",
            "Favorite Films",
        ),
        [
            (
                "2015-03-01",
                "owner",
                "Given",
                "Family",
                EMAIL,
                "Somewhere",
                "https://example.com",
                "A bio nobody asked for",
                "they/them",
                ", ".join(favorites),
            )
        ],
    )


def _discarded() -> dict[str, str]:
    """Everything else the archive carries. Reading any of it is a bug this catches."""
    unread = Row("Never Read This", 1999)
    return {
        "reviews.csv": _csv(
            (
                "Date",
                "Name",
                "Year",
                "Letterboxd URI",
                "Rating",
                "Rewatch",
                "Review",
                "Tags",
                "Watched Date",
            ),
            [
                (
                    unread.date,
                    unread.name,
                    "1999",
                    unread.uri_text,
                    "5",
                    "",
                    "words",
                    "",
                    unread.date,
                )
            ],
        ),
        "comments.csv": _csv(("Date", "Content", "Comment"), [("2024-01-01", "a list", "a word")]),
        "likes/films.csv": _csv(
            ("Date", "Name", "Year", "Letterboxd URI"),
            [(unread.date, unread.name, "1999", unread.uri_text)],
        ),
        # The two folders whose diary.csv is byte-identical in shape to the real one:
        # anything matching on the file name alone rather than its place in the archive
        # imports deleted and orphaned rows as though the owner had watched them.
        "deleted/diary.csv": _diary((unread,)),
        "orphaned/diary.csv": _diary((unread,)),
    }


def _csv(header: tuple[str, ...], rows: list[tuple[str, ...]]) -> str:
    out = io.StringIO(newline="")
    writer = csv.writer(out, lineterminator="\n")
    writer.writerow(header)
    writer.writerows(rows)
    return out.getvalue()


def _zip(members: dict[str, str]) -> bytes:
    buffer = io.BytesIO()
    with zipfile.ZipFile(buffer, "w", zipfile.ZIP_DEFLATED) as archive:
        for name, body in members.items():
            archive.writestr(name, body)
    return buffer.getvalue()
